metronome/checkpointing/interpreter/src/io/iohk/metronome/checkpointing/interpreter/messages/InterpreterMessage.scala

```diff
@@ -0,0 +1,147 @@
+    * the next view when the replica becomes leader. In
+    * that round, the Service should send a `GetCheckpointCandidateRequest`.
+    *
+    * This is an optimisation, so we don't send the `Ledger` in
+    * futile attempts when there's no chance for a candidate to
+    * be produced.
+    */
+  case class NewCheckpointCandidateRequest(
+      requestId: RequestId
+  ) extends InterpreterMessage
+      with Request
+      with FromInterpreter
+      with NoResponse
+
+  /** When it becomes a leader of a view, the Service asks
+    * the Interpreter to produce a new checkpoint candidate,
+    * given the current state of the ledger.
+    *
+    * A response is expected even when there's no candidate
+    * to put in a block, so that we can move on to the next
+    * leader after an idle round, but without a timeout.
+    */
+  case class GetCheckpointCandidateRequest(
+      requestId: RequestId,
+      ledger: Ledger
+  ) extends InterpreterMessage
+      with Request
+      with FromService
+
+  /** The Interpreter may or may not be able to produce a new
+    * checkpoint candidate, depending on whether the conditions
+    * are right (e.g. the next checkpointing height has been reached).
+    *
+    * The response should contain `None` if we should just move on
+    * to the next leader, or `Some` if a block is to be prepared.
+    */
+  case class GetCheckpointCandidateResponse(
+      requestId: RequestId,
+      maybeCheckpointCandidate: Option[Transaction.CheckpointCandidate]
+  ) extends InterpreterMessage
+      with Response
+      with FromInterpreter
+
+  /** The Service asks the Interpreter to validate all transactions
+    * in a block, given the current ledger state, and return the
+    * updated ledger.
+    *
+    * This could be done transaction by transaction, but that would
+    * require sending the ledger every step along the way, which
+    * would be less efficient.
+    *
+    * The updated ledger is returned so the interpeter can decide
+    * which proposer blocks to keep after applying any checkpoints.
+    *
+    * If the Interpreter doesn't have enough data to validate the
+    * block, it should hold on to it until it does, only responding
+    * when it has the final conclusion.
+    */
+  case class ValidateBlockRequest(
+      requestId: RequestId,
+      blockBody: Block.Body,
+      ledger: Ledger
+  ) extends InterpreterMessage
+      with Request
+      with FromService
+
+  /** The Interpreter responds to the block validation request when
+    * it has all the data available to perform the validation.
+    *
+    * If the block was valid, `Some` updated ledger is returned,
+    * otherwise the response contains `None`.
+    *
+    * The Service will check whether the hash of the updated ledger
+    * matches the `postStateHash` in the block header.
+    */
+  case class ValidateBlockResponse(
+      requestId: RequestId,
+      maybeUpdatedLedger: Option[Ledger]
```

Comment:
Here we could use a `result: Either[Error, Ledger]` if there is something to be communicated back about the reason for failure.